src/coreclr/vm/codeman.cpp

```diff
@@ -3713,16 +3822,44 @@ void EECodeGenManager::FreeCodeMemory(HostCodeHeap *pCodeHeap, void * codeStart)
     }
     CONTRACTL_END;
 
-    CrstHolder ch(&m_CodeHeapCritSec);
+    CrstHolder ch(&m_CodeHeapLock);
+    if (m_iteratorCount == 0)
+    {
+        FreeHostCodeHeapMemoryWorker(pCodeHeap, codeStart);
+    }
+    else
+    {
+        // If we are in the middle of an enumeration, we cannot unload any code heap code.
+        EX_TRY
+        {
+            FreeHostCodeHeapStartCode* delayed = m_delayFreeHostCodeHeapMemory.AppendThrowing();
```

Comment:
Is the code header going to point to DynamicMethodDesc that may get reused for other methods? Can it cause problems?

We may want to delay freeing the whole DynamicMethodDescs.

Reply:
This is done.